### PR TITLE
RSDK-5535: Implement console formatting for logger.

### PR DESCRIPTION
**Pull request by dgottlieb**

### CODE REVIEW

**Review by maximpertsov — CHANGES_REQUESTED** [review]
I found some places where we use `fmt.Sprint[f]` on arguments that are all already stringified, but otherwise looks good.

File: logging/impl.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+	return nil
+}
+
+func (impl *impl) With(args ...interface{}) *zap.SugaredLogger {
+	return nil
+}
+
+func (impl *impl) WithOptions(opts ...zap.Option) *zap.SugaredLogger {
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (dgottlieb) [reply logging/impl.go:45]: That's doubtful for the functions returning a `SugaredLogger` as they have limited utility. Right now these are here to satisfy the `Logger` interface. For now I'm punting on what to do here until I actually start switching over and seeing what our options are. There aren't many usages of these

File: logging/impl.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+	const prefixLength = 6
+	toPrint := make([]any, prefixLength)
+	toPrint[0] = time.Now().Format("2006-01-02T15:04:05.000Z0700")
+	toPrint[1] = "\t"
+	toPrint[2] = strings.ToUpper(logLevel.String())
+	toPrint[3] = "\t"
+	toPrint[4] = getCaller()
+	toPrint[5] = "\t"
+
+	return fmt.Sprint(toPrint...)
```
[the remark this answers is not in the document]
Comment (dgottlieb) [reply logging/impl.go:65]: Oh, this is a good idea! Given this isn't being used, I'd prefer to take the scientific approach of writing benchmarks to see the effect each of these individual changes has. Even if the change is small and a benefit is expected!

I've [filed a ticket](https://viam.atlassian.net/browse/RSDK-5568) about benchmarking/optimization. There are some other optimizations worth trying (for code not yet written). Would you be ok with me postponing this to later in the project?

File: logging/impl.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+}
+
+// Forwards to `fmt.Sprintf` which does standard Go formatting.
+func (impl *impl) formatf(logLevel Level, template string, args ...interface{}) string {
+	toPrint := impl.getPrefix(logLevel)
+	logValue := fmt.Sprintf(template, args...)
+
+	return fmt.Sprintf("%s%s", toPrint, logValue)
```
[the remark this answers is not in the document]
Comment (dgottlieb) [reply logging/impl.go:85]: Noted in the ticket.

File: logging/impl.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+	var encMapStr string
+	encMapBytes, err := json.Marshal(mp)
+	if err == nil {
+		encMapStr = string(encMapBytes)
+	} else {
+		encMapStr = fmt.Sprintf("%v", mp)
+	}
+
+	// The prefix has a trailing tab character. Directly concatenate the `prefix` with the `msg`.
+	return fmt.Sprintf("%s%s\t%s", prefix, msg, encMapStr)
```
[the remark this answers is not in the document]
Comment (dgottlieb) [reply logging/impl.go:118]: Noted in the ticket.

**Comment by dgottlieb** [comment]
Right now I'm working on the test failures. It's passing locally. I think* the test machines are configured UTC which is dropping most of the `-0400` timezone information.

**Comment by dgottlieb** [comment]
Yeah, switching to GMT causes the same failure locally:
```
$ TZ="/usr/share/zoneinfo/GMT" V go test -run=TestConsoleOutputFormat
2023-10-31T13:12:09.240Z	INFO	logging/impl_test.go:137	impl Info log

--- FAIL: TestConsoleOutputFormat (0.00s)
    impl_test.go:140: Expected: '28'
        Actual:   '24'
        (Should be equal)
FAIL
exit status 1
FAIL	go.viam.com/rdk/logging	0.002s
```

**Comment by dgottlieb** [comment]
An example of the new logging output:
```
=== RUN   TestConsoleOutputFormat
    impl_test.go:144: Expected: '24'
        Actual:   '28'
        (Should be equal)
        Message: Date length mismatch
        Log line: `2023-10-31T09:39:21.059-3900	INFO	logging/impl_test.go:141	impl Info log`
--- FAIL: TestConsoleOutputFormat (0.00s)
FAIL
FAIL	go.viam.com/rdk/logging	0.010s
FAIL
```

**Comment by dgottlieb** [comment]
Actually, pause the review for a sec. This got refactored more than I thought it would with the next ticket.

**Review by maximpertsov — COMMENTED** [review]
Looks good! However, I'm unsure why the tests are failing on some platforms

**Comment by dgottlieb** [comment]
> Looks good! However, I'm unsure why the tests are failing on some platforms

The 32-bit tests hit a timeout on the motion/builtin tests:
```
goroutine 1 [chan receive, 8 minutes]:
testing.tRunner.func1()
	/usr/lib/go-1.19/src/testing/testing.go:1412 +0x510
testing.tRunner(0x4371a40, 0x46eddac)
	/usr/lib/go-1.19/src/testing/testing.go:1452 +0x160
testing.runTests(0x4cc1a60, {0x34797e8, 0x13, 0x13}, {0xc148708d04d4d027, 0x8bcd44e85c, 0x3a0b580})
	/usr/lib/go-1.19/src/testing/testing.go:1844 +0x3e8
testing.(*M).Run(0x4421220)
	/usr/lib/go-1.19/src/testing/testing.go:1726 +0x5f0
go.viam.com/utils/testutils/ext.VerifyTestMain({0x24eb528, 0x4421220})
	/home/testbot/go/pkg/mod/go.viam.com/utils@v0.1.52/testutils/ext/verify.go:52 +0xf8
go.viam.com/rdk/services/motion/builtin.TestMain(...)
	/rdk/services/motion/builtin/verify_main_test.go:11
```

In the arm tests, the motion tests had a test failure:
```
    logger.go:130: 2023-10-31T18:39:46.441Z	DEBUG	builtin/move_request.go:111	deviation from plan: %v(0.000000000000000000000000, 0.000000000000000000000000, 0.000000000000000000000000)
    logger.go:130: 2023-10-31T18:39:46.441Z	DEBUG	builtin/builtin.go:355	execution response: builtin.moveResponse{success: true, err: <nil>}
    builtin_test.go:1323: Expected: false
        Actual:   true
    --- FAIL: TestObstacleDetection/ensure_replan_due_to_obstacle_collision (13.47s)

=== FAIL: services/motion/builtin TestObstacleDetection (0.00s)
```

The logging test hasn't failed since I changed our time format to always be UTC (instead of depending on the machine/OS timezone setting).

**Review by maximpertsov — APPROVED** [review]

**Pull request merged**